### PR TITLE
Fix ios build error when use_frameworks is on and fabric is off (v2)

### PR DESCRIPTION
**Pull request by Kudo**
## Summary

alternative solution for #33379 

> when `use_frameworks!` is on, there are errors like:
> ```
> 'FBReactNativeSpec/FBReactNativeSpec.h' file not found
> #import <FBReactNativeSpec/FBReactNativeSpec.h>
> ```
> this error may come from from https://github.com/facebook/react-native/commit/f7e4c07c84b6 regression.
> 
> when `use_frameworks!` is on, xcode will search headers from framework directories, the correct imports would be `#import <React_Codegen/FBReactNativeSpec/FBReactNativeSpec.h>` (xcode will transform dash to underscore, so it is `React_Codegen` but not `React-Codegen`). in the other hand, when `use_frameworks!` is off, the correct import is `#import <React-Codegen/FBReactNativeSpec/FBReactNativeSpec.h>`.
>
> 
> this fix is specific for old architecture (fabric is off).
> 
> when fabric is on, there are other errors from duplicated headers when copying to build folder. [the reason is that framework build would try to flatten headers](https://mkonrad.net/2015/03/29/xcode-static-libraries-preserving-header-directory-structure.html). we have `primitives.h` in different folders and they would be flattened into `React_Fabric.framework/Headers`. to be honest, i don't know how to deal with the problem in the meantime,  maybe subspecs are not enough, we should separate them from subspecs to dedicated podspecs so that we can have these targets as different frameworks.

in this alternative fix, i try to add `React-Codegen/React_Codegen.framework/Headers` into header search paths and make original `#import <FBReactNativeSpec/FBReactNativeSpec.h>` reachable.

[this change](https://github.com/facebook/react-native/commit/7a0398c331f22abc619a64b444ec7153357b0a30) in the pr is just a workaround to solve breaking in latest main branch and this is not important to the `use_frameworks!` fix at all. this breaking was coming from https://github.com/facebook/react-native/commit/180495159517dc0bfa103621e5ff62fc04cb3c8b. 

## Changelog

[iOS] [Fixed] - Fix iOS build error when Podfile `use_frameworks!` is on and Fabric is off

## Test Plan

verify with rn-tester
1. change `fabric_enabled` to false in `packages/rn-tester/Podfile`
2. `USE_FRAMEWORKS=1 pod install`
3. build rn-tester in xcode

### CODE REVIEW

**Comment by Kudo** [comment]
@mikehardy could you help to verify this solution for your cases when you get a chance?
this is the patch file for v0.68.0-rc.2: https://gist.github.com/Kudo/7a590cbd6f8ca33fdf9f706ac96049ea

**Comment by cortinico** [comment]
It's all green interally so we should be good to go 👍

**Comment by mikehardy** [comment]
Fantastic! Verifying now @Kudo

**Comment by mikehardy** [comment]
It appears to work, applying this patch https://gist.githubusercontent.com/Kudo/7a590cbd6f8ca33fdf9f706ac96049ea/raw/00ff89b8c56b30f0db90bf1c74d4cdbe7c9f1612/react-native+0.68.0-rc.2.patch during my normal test of static frameworks build on iOS resulted in successful compilation + execution of the app

Nice!

File: React/CoreModules/RCTExceptionsManager.h [hunk 1]
```diff
@@ -37,7 +36,7 @@ NS_ASSUME_NONNULL_BEGIN
 - (void)reportFatalException:(nullable NSString *)message
                        stack:(nullable NSArray<NSDictionary *> *)stack
                  exceptionId:(double)exceptionId;
-- (void)reportEarlyJsException:(std::string)errorMap;
+- (void)reportEarlyJsException:(nullable NSString *)errorMap;
```
Comment (dmitryrykun) [inline React/CoreModules/RCTExceptionsManager.h:39]: @Kudo please revert this change. It is not related to the original issue, and it breaks our internal tests.
Other than that, this solution looks good!

**Review by dmitryrykun — CHANGES_REQUESTED** [review]

File: React/CoreModules/RCTExceptionsManager.mm [hunk 2]
```diff
@@ -138,9 +138,9 @@ - (void)reportFatal:(NSString *)message stack:(NSArray<NSDictionary *> *)stack e
   }
 }
 
-- (void)reportEarlyJsException:(std::string)errorMap
+- (void)reportEarlyJsException:(NSString *)errorMap
 {
-  NSString *errprStr = [NSString stringWithUTF8String:errorMap.c_str()];
+  NSString *errprStr = errorMap;
```
Comment (mikehardy) [inline React/CoreModules/RCTExceptionsManager.mm:143]: would these also need revert @dmitryrykun ? just scanning this after seeing your cmoment and the seem similar

Comment (dmitryrykun) [reply React/CoreModules/RCTExceptionsManager.mm:143]: Yeah, I meant all the changes in RCTExceptionsManager.

**Comment by Kudo** [comment]
reverted the commit. thanks @dmitryrykun for review and @mikehardy for kindly help to verify the fix 💛

**Comment by cortinico** [comment]
Thank you all for the great work addressing this issue 🙏
